addons/binding/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/internal/convention/homeassistant/HaID.java

```diff
@@ -64,62 +68,123 @@ public HaID(String mqttTopic) {
-        String groupId = channel.getGroupId();
-        if (groupId == null) {
-            throw new IllegalArgumentException("Channel needs a group ID!");
+    public static HaID fromConfig(String baseTopic, Configuration config) {
+        String objectID = (String) config.get("objectid");
+        String nodeID = (String) config.getProperties().getOrDefault("nodeid", "");
+        String component = (String) config.get("component");
+        return new HaID(baseTopic, objectID, nodeID, component);
+    }
+
+    public void toConfig(Configuration config) {
+        config.put("objectid", objectID);
+        config.put("nodeid", nodeID);
+        config.put("component", component);
+    }
+
+    public HandlerConfiguration toHandlerConfiguration() {
+        String objectID = this.objectID;
+        if (StringUtils.isNotBlank(nodeID)) {
+            objectID = nodeID + "/" + objectID;
+        }
+
+        return new HandlerConfiguration(baseTopic, objectID);
+    }
+
+    public static HaID fromConfig(HandlerConfiguration config) {
+        String baseTopic = config.getBasetopic();
+        String objectID = config.getObjectid();
+        String nodeID = "";
+
+        if (StringUtils.contains(objectID, '/')) {
+            String[] parts = objectID.split("/");
+
+            if (parts.length != 2) {
+                throw new IllegalArgumentException(
+                        "Bad configuration. objectid must be <objectId> or <nodeId>/<objectId>!");
+            }
+            nodeID = parts[0];
+            objectID = parts[1];
         }
-        String[] groupParts = groupId.split("_");
-        if (groupParts.length != 2) {
-            throw new IllegalArgumentException("Channel needs a group ID with the pattern component_node!");
+        return new HaID(baseTopic, objectID, nodeID, "+");
+    }
+
+    public String getFallbackGroupId() {
+        StringBuilder str = new StringBuilder();
+
+        if (StringUtils.isNotBlank(nodeID)) {
+            str.append(nodeID).append('_');
         }
-        this.objectID = channel.getThingUID().getId();
-        this.nodeID = groupParts[1];
-        this.component = groupParts[0];
-        this.baseTopic = baseTopic;
+        str.append(objectID).append('_').append(component);
+        return str.toString();
     }
 
-    /**
-     * We map the HomeAssistant MQTT topic tree object to an ESH Thing.
-     */
-    public String getThingID() {
+    public String getComponent() {
+        return component;
+    }
+
+    public String getObjectID() {
         return objectID;
     }
 
-    /**
-     * The channel group type UID consists of all components of this object (object-id + node-id + component-id).
-     */
-    public String getChannelGroupTypeID() {
-        return objectID + "_" + component + nodeID;
+    public String getTopic(String suffix) {
+        StringBuilder str = new StringBuilder();
+
+        str.append(baseTopic).append('/').append(component).append('/');
```

Comment:
Please precompute that and store it in a private field and then just do a string concatenation: `_stored_var+suffix`.

Reply:
I am not sure about this one.
This is a trde off between execution time and space.
While the mehtod is called max. a couple of times during initialization, the member will take up space for the whole runtime....

Reply:
Yup a tradeoff. It's up to you. Maybe other fields are not used and can be dumped?